assets/src/stories-editor/blocks/amp-story-text/edit.js

```diff
@@ -25,73 +24,82 @@ import { getBackgroundColorWithOpacity } from '../../../common/helpers';
 import './edit.css';
 import { DraggableText } from '../../components';
 
-class TextBlockEdit extends Component {
-	constructor( ...args ) {
-		super( ...args );
-
-		this.state = {
-			isEditing: false,
-			hasOverlay: true,
-		};
-	}
-
-	componentDidMount() {
-		maybeUpdateFontSize( this.props );
-	}
-
-	componentDidUpdate( prevProps, prevState ) {
-		const { attributes, fontSize, isSelected } = this.props;
-		const {
-			height,
-			width,
-			content,
-			ampFitText,
-			ampFontFamily,
-		} = attributes;
-
-		// If the block was unselected, make sure that it's not editing anymore.
-		if ( ! isSelected && prevProps.isSelected ) {
-			this.toggleIsEditing( false );
-			this.toggleOverlay( true );
+const TextBlockEdit = ( props ) => {
+	const {
+		attributes,
+		setAttributes,
+		className,
+		clientId,
+		fontSize,
+		isPartOfMultiSelection,
+		isSelected,
+		backgroundColor,
+		customBackgroundColor,
+		textColor,
+		name,
+	} = props;
+	const {
+		placeholder,
+		content,
+		align,
+		ampFontFamily,
+		ampFitText,
+		autoFontSize,
+		width,
+		height,
+		opacity,
+	} = attributes;
+
+	const [ isEditing, setIsEditing ] = useState( false );
+	const [ hasOverlay, setHasOverlay ] = useState( true );
+
+	const toggleIsEditing = ( enable ) => {
+		if ( enable !== isEditing ) {
+			setIsEditing( ! isEditing );
 		}
+	};
 
-		const checkFontSize = ampFitText && (
-			prevProps.attributes.ampFitText !== ampFitText ||
-			prevProps.attributes.ampFontFamily !== ampFontFamily ||
-			prevProps.attributes.width !== width ||
-			prevProps.attributes.height !== height ||
-			prevProps.attributes.content !== content
-		);
+	const toggleOverlay = ( add ) => {
+		if ( add !== hasOverlay ) {
+			setHasOverlay( ! hasOverlay );
+		}
+	};
 
-		if ( checkFontSize ) {
-			maybeUpdateFontSize( this.props );
+	useEffect( () => {
+		if ( isEditing ) {
+			setInputSelectionToEnd( '.is-selected .wp-block-amp-amp-story-text' );
 		}
+	}, [ isEditing ] );
 
-		const checkBlockDimensions = ! ampFitText && (
-			! isEqual( prevProps.fontSize, fontSize ) ||
-			prevProps.attributes.ampFitText !== ampFitText ||
-			prevProps.attributes.ampFontFamily !== ampFontFamily ||
-			prevProps.attributes.content !== content
-		);
+	useEffect( () => {
+		if ( ampFitText ) {
+			maybeUpdateFontSize( props );
+		}
+	}, [ ampFitText, ampFontFamily, width, height, content ] );
```

Comment:
This is super messy, because the "helper" is stateful and uses external information making it hard to correctly guard for when it needs to re-run - also it has side-effects. What if this method changes internally and this guard isn't updated?

The helper should received all arguments directly, be stateless and be side-effect free for this to be "proper".

I'm not a big fan of this, but it's probably too huge a task to rewrite these helpers too.

Reply:
You mean the helper should receive all arguments individually instead of just  `props`?